File: kotlin/day01/sonarSweep.kt
package day01

import java.io.File

fun main() {
    val depths = parseInput()
    val part1 = countIncreases(depths)
    val slidingWindows = getSlidingWindows(depths)
    val part2 = countIncreases(slidingWindows)

    println(part1)
    println(part2)
}

fun parseInput(): List<Int> {
    return File("../../input/01.txt")
        .readLines()
        .map { it.toInt() }
}

fun countIncreases(measurements: List<Int>): Int {
    return (measurements.dropLast(1) zip measurements.drop(1))
        .count { it.first < it.second }
}

fun getSlidingWindows(depths: List<Int>): List<Int> {
    return zipSum(
        zipSum(
            depths.dropLast(2),
            depths.dropLast(1).drop(1)
        ),
        depths.drop(2)
    )
}

fun zipSum(first: List<Int>, second: List<Int>): List<Int> {
    return (first zip second).map{ it.first + it.second }
}

File: python/day01/sonar_sweep.py
INPUT_FILE = "../../input/01.txt"


def parse_input() -> list[int]:
    """
    Parses the input and returns a list of depth measurements.
    """
    with open(INPUT_FILE, "r") as f:
        return [int(line) for line in f]


def count_increases(measurements: list[int]) -> int:
    """
    Counts the number of times a measurement increases from the previous measurement.
    """
    return sum(1 for a, b in zip(measurements[:-1], measurements[1:]) if a < b)


def get_sliding_windows(depths: list[int]) -> list[int]:
    """
    Calculates the three-measurement sliding windows.
    """
    return [a + b + c for a, b, c in zip(depths[:-2], depths[1:-1], depths[2:])]


if __name__ == "__main__":
    depths = parse_input()
    part1 = count_increases(depths)
    sliding_windows = get_sliding_windows(depths)
    part2 = count_increases(sliding_windows)

    print(part1)
    print(part2)
